_posts/2023-11-01-#28368.md

```diff
@@ -7,19 +7,61 @@ authors: [ismaelsadeeq]
 host: ismaelsadeeq
 status: upcoming
-commit:
+commit: 79bcc5ca0679daf6e57fc4d7ce2244262a7cfd13
 ---
 
-_Notes and questions to follow soon!_
 
-<!-- TODO: Before meeting, add notes and questions
 ## Notes
 
+* In this week's review club, we will discuss bitcoin core PR #28368 which modifies the fee estimator `CBlockPolicyEstimator` to update from the `CValidationInterface` / `CScheduler` thread notifications instead of mempool.
+
+* The [`CBlockPolicyEstimator`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/policy/fees.h#L146C13-L146C13) is used to estimate fees based on a confirmation target. We have two RPCs, [`estimaterawfees`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/rpc/fees.cpp#L100) and [`estimatesmartfees`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/rpc/fees.cpp#L29) used to obtain a fee estimate for a given confirmation target, it is also heavily used in the [wallet](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/wallet/fees.cpp#L29).
+
+* Before this PR the [`CBlockPolicyEstimator`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/policy/fees.h#L146C13-L146C13) is coupled to the mempool such that whenever a transaction is added or removed from the mempool in [`addUnchecked`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L475) and [`removedUnchecked`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L787) methods we also update the fee estimator.
+
+* Similarly, whenever a new block is connected, the mempool in the [`removeForBlock`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L487) method removes the block transactions that are in the mempool together with its ancestors, this also triggers an update to the fee estimator.
+ 
+* This update is slowing down [`removeForBlock`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L487) because the mempool [`cs`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L391C1-L391C1) is held and waits for the fee estimator to finish updating before getting out of the [`removeForBlock`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L487) scope, hence blocking [`ConnectTip`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/validation.h#L732C13-L732C13), adding complex steps to the fee estimator during this update can be inefficient.
+
+* The [`CValidationInterface`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/validationinterface.h#L77) is a class any subscriber can implement to receive event callbacks whenever they are generated.
```

Comment:
I wouldn't really say callbacks are "received" or "emitted". I'd maybe say that validation "fires signals" or "emits events" and a client that *registered* a callback "receives a notification." A thread *executes* a callback.

Do you want to mention that some of these callbacks are run *asynchronously* on a different thread from the one that fires the event? That's the key benefit of this PR. A few other notes might help as well, as I think it's probably the biggest concept for new reviewers to learn for this PR.

Reply:
Reword it to your suggestion, and explicitly state that the fee estimator now updates asynchronously in the background